fastruby/points
# Handle multiple estimations for one user and N+1 queries

**Description:**

This PR fixes 2 issues. I started working on #141 but some changes were too related to #136.

There's currently a bug that allows estimators to add more than 1 estimate for a given story (#142), even though when that bug gets fixed, we still need to handle old data that's already in the DB and this PR does that.

For 141, I refactored some methods to always get the first estimate by a given user and only counting 1 estimate per user when doing totals and averages.

These changes were closely related to 136 because to optimize the way the refactor works I had to add a bunch of `includes(:estimates)` in some places, which removes the N+` queries.

**How to test/QA:**

You can follow the steps in #142 to generate more than 1 estimate for a user and then you can check in the report that the totals and averages won't match the actual total and average of the table before this fix.

To test #136, you can check the logs when you open a project's show action, now the logs look like:

```
  Project Load (0.1ms)  SELECT "projects".* FROM "projects" WHERE "projects"."id" = $1 LIMIT $2  [["id", 2], ["LIMIT", 1]]
  Rendered shared/_project_title.html.erb (Duration: 1.6ms | Allocations: 1659)
  Story Load (0.3ms)  SELECT "stories".* FROM "stories" WHERE "stories"."project_id" = $1 ORDER BY position ASC NULLS FIRST, created_at ASC  [["project_id", 4]]
  Estimate Load (0.1ms)  SELECT "estimates".* FROM "estimates" WHERE "estimates"."story_id" IN ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11)  [["story_id", 9], ["story_id", 23], ["story_id", 24], ["story_id", 25], ["story_id", 26], ["story_id", 27], ["story_id", 28], ["story_id", 29], ["story_id", 30], ["story_id", 31], ["story_id", 32]]
  User Load (0.1ms)  SELECT "users".* FROM "users" WHERE "users"."id" = $1 LIMIT $2  [["id", 3], ["LIMIT", 1]]
  Story Load (0.1ms)  SELECT "stories".* FROM "stories" WHERE "stories"."project_id" = $1  [["project_id", 4]]
  CACHE Estimate Load (0.0ms)  SELECT "estimates".* FROM "estimates" WHERE "estimates"."story_id" IN ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11)  [["story_id", 9], ["story_id", 23], ["story_id", 24], ["story_id", 25], ["story_id", 26], ["story_id", 27], ["story_id", 28], ["story_id", 29], ["story_id", 30], ["story_id", 31], ["story_id", 32]]
  CACHE User Load (0.0ms)  SELECT "users".* FROM "users" WHERE "users"."id" = $1 LIMIT $2  [["id", 3], ["LIMIT", 1]]
  CACHE Story Load (0.0ms)  SELECT "stories".* FROM "stories" WHERE "stories"."project_id" = $1  [["project_id", 4]]
  CACHE Estimate Load (0.0ms)  SELECT "estimates".* FROM "estimates" WHERE "estimates"."story_id" IN ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11)  [["story_id", 9], ["story_id", 23], ["story_id", 24], ["story_id", 25], ["story_id", 26], ["story_id", 27], ["story_id", 28], ["story_id", 29], ["story_id", 30], ["story_id", 31], ["story_id", 32]]
  CACHE User Load (0.0ms)  SELECT "users".* FROM "users" WHERE "users"."id" = $1 LIMIT $2  [["id", 3], ["LIMIT", 1]]
```
with only 3 queries (project, stories, estimate) and some cached queries that are 0.0ms and no N+1

Closes #141 
Closes #136 

___

I will abide by the [code of conduct](https://github.com/fastruby/points/blob/main/pull_request_template.md).
